packages/hierarchy-builder/src/hierarchy-builder/TreeNodesProvider.ts

```diff
@@ -0,0 +1,515 @@
+      createMergeInstanceNodesByLabelReducer(this._directNodesCache),
+      createHideIfNoChildrenReducer((n) => this.hasNodes(n), false),
+      createHideNodesInHierarchyReducer(this, this._directNodesCache, false),
+      sortNodesByLabelReducer,
+      createClassGroupingReducer(this._schemas),
+    );
+    return parentNode ? result.pipe(createPersistChildrenReducer(parentNode)) : result;
+  }
+
+  public async getNodes(parentNode: TreeNode | undefined): Promise<TreeNode[]> {
+    return new Promise((resolve, reject) => {
+      const nodes = new Array<TreeNode>();
+      this.getNodesObservable(parentNode)
+        // finalize before returning
+        .pipe(createDetermineChildrenReducer(this), supplyIconsReducer)
+        // load all nodes into the array and resolve
+        .subscribe({
+          next(node) {
+            nodes.push(node);
+          },
+          error(err) {
+            reject(err);
+          },
+          complete() {
+            resolve(nodes);
+          },
+        });
+    });
+  }
+
+  public hasNodes(node: TreeNode): Observable<boolean> {
+    const enableLogging = false;
+    if (Array.isArray(node.children)) {
+      return of(node.children.length > 0);
+    }
+
+    const directChildren = this.ensureDirectChildren(node);
+    return directChildren
+      .pipe(
+        tap((n) => enableLogging && console.log(`HasNodes: partial node before HideIfNoChildrenReducer: ${JSON.stringify(n)}`)),
+        createHideIfNoChildrenReducer((n) => this.hasNodes(n), true),
+        tap((n) => enableLogging && console.log(`HasNodes: partial node before HideNodesInHierarchyReducer: ${JSON.stringify(n)}`)),
+        createHideNodesInHierarchyReducer(this, this._directNodesCache, true),
+      )
+      .pipe(
+        tap((n) => enableLogging && console.log(`HasNodes: partial node before mapping to 'true': ${JSON.stringify(n)}`)),
+        map(() => true),
+        take(1),
+        defaultIfEmpty(false),
+        catchError((e: Error) => {
+          enableLogging && console.log(`HasNodes: error while determining children: ${e.message}`);
+          if (e.message === "rows limit exceeded") {
+            return of(true);
+          }
+          throw e;
+        }),
+        tap((r) => enableLogging && console.log(`HasNodes: result: ${r}`)),
+      );
+  }
+}
+
+function createPersistChildrenReducer(parentNode: TreeNode) {
+  return function (nodes: Observable<InProgressTreeNode>): Observable<InProgressTreeNode> {
+    nodes.pipe(
+      reduce((acc, node) => [...acc, node], new Array<InProgressTreeNode>()),
+      tap((list) => {
+        if (Object.isExtensible(parentNode)) {
+          parentNode.children = list;
+        } else {
+          // eslint-disable-next-line no-console
+          console.log(`node ${parentNode.label} not extensible for setting children`);
+        }
+      }),
+    );
+    return nodes;
+  };
+}
+
+function createDetermineChildrenReducer(provider: TreeNodesProvider) {
```

Comment:
Maybe this function could take a callback instead of `provider`? Same as `createHideIfNoChildrenReducer`

Reply:
Good point, changed.